# Part 1:

In [1]:
# Import data libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path = '/Users/vinhdao/Documents/Instacart Basket Analysis'

In [3]:
# Import new customer data

customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

## Step 3: Checking the dataset

In [4]:
# Checking the data set

customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
customers.shape

(206209, 10)

In [7]:
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


## Step 4: Wrangling the dataset

In [19]:
# Rename First Name as first_name

customers.rename(columns = {'First Name' : 'First_Name'}, inplace = True)

In [10]:
# Rename Surnam as surname

customers.rename(columns = {'Surnam' : 'Surname'}, inplace = True)

In [12]:
# Change the capital of STATE become State

customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [13]:
# Rename fam_status as marital_status

customers.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

## Step 5: Checking data quality and consistency

In [49]:
# Descriptive analysis

customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [15]:
# Checking the missing data

customers.isnull().sum()

user_id               0
first_name        11259
surname               0
Gender                0
State                 0
Age                   0
date_joined           0
n_dependants          0
marital_status        0
income                0
dtype: int64

As the result suggested, there are 11259 value missing the first name of the customers. This is not the key values, so we don't need to address more. Howevers, for further learning for this, we will go further deeper for the understanding

In [21]:
df_nan = customers[customers['first_name'].isnull() == True]

In [22]:
df_nan

user_id first_name   surname  Gender         State  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants marital_status  income  
53                 2        married   41709  
73                 0         single   82518  
82                 3        married  117099  
99                 1        married  155673  
105                3        married  151819  
...              ...            ...     ...  
206038             3        married   87783  
206044             2        married  108488  
206090             1        married   45275  
206162             0         single   56325  
206171             2        married   59222  

[11259 rows x 10 columns]

Because these are the name of customers, we use any method to addressing the missing values. For this, we will leave it 

In [25]:
# Checking duplicates

df_dups = customers[customers.duplicated()]

In [24]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, surname, Gender, State, Age, date_joined, n_dependants, marital_status, income]
Index: []

As the result suggested, there are no duplicated in the dataset

In [26]:
df_test = pd.DataFrame()

In [29]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [30]:
# Let's check again the data type
customers.dtypes

user_id            int64
first_name        object
surname           object
Gender            object
State             object
Age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

From above, we can see that all the data types is consistent and no need to change

## Step 6: Combine the dataset

In [47]:
# Import the rest of prepared Instacart data

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merge.pkl'))

In [44]:
ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   
1   1977647      138    prior            30                   6  ...   
2    389851      709    prior             2                   0  ...   
3    652770      764    prior             1                   3  ...   
4   1813452      764    prior             3                   4  ...   
5   1701441      777    prior            16                   1  ...   
6   1871483      825    prior             3                   2  ...   
7   1290456      910    prior            12                   3  ...   
8    369558     1052    prior            10                   1  ...   
9    589712     1052    prior            15                   1  ...   

   price_range_loc   prices_range_loc     busiest_day  busiest_period_of_day  \
0              NaN  Mid-range product  Regularly busy            Most orders   
1              NaN  Mid-range product  Regularly busy         Average orders   
2              NaN  Mid-range product    Busiest days         Average orders   
3              NaN  Mid-range product    Slowest days            Most orders   
4              NaN  Mid-range product    Slowest days         Average orders   
5              NaN  Mid-range product    Busiest days         Average orders   
6              NaN  Mid-range product  Regularly busy            Most orders   
7              NaN  Mid-range product    Slowest days            Most orders   
8              NaN  Mid-range product    Busiest days         Average orders   
9              NaN  Mid-range product    Busiest days            Most orders   

  max_order      loyalty_flag average_spend spender_flag Customer_frequency  \
0        32  Regular customer      6.935811  Low_spender                8.0   
1        32  Regular customer      6.935811  Low_spender                8.0   
2         5      New customer      7.930208  Low_spender                8.0   
3         3      New customer      4.972414  Low_spender                9.0   
4         3      New customer      4.972414  Low_spender                9.0   
5        26  Regular customer      6.935398  Low_spender               11.0   
6         9      New customer      5.957576  Low_spender               20.0   
7        12  Regular customer      6.680000  Low_spender                6.0   
8        20  Regular customer      7.162500  Low_spender               10.0   
9        20  Regular customer      7.162500  Low_spender               10.0   

      frequency_flag  
0  Frequent Customer  
1  Frequent Customer  
2  Frequent Customer  
3  Frequent Customer  
4  Frequent Customer  
5   Regular customer  
6   Regular customer  
7  Frequent Customer  
8  Frequent Customer  
9  Frequent Customer  

[10 rows x 25 columns]

In [36]:
ords_prods_merge.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                      int64
eval_set                    object
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
prices_range_loc            object
busiest_day                 object
busiest_period_of_day       object
max_order                    int64
average_spend              float64
spender_flag                object
Customer_frequency         float64
frequency_flag              object
dtype: object

In [48]:
ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [50]:
# Merge the dataset

ords_prods_customers = ords_prods_merge.merge(customers, on = 'user_id', indicator = True)

In [51]:
# Export the dataset

ords_prods_customers.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merge.pkl'))